src/object/cli_ec.c

```diff
@@ -334,31 +423,50 @@ ec_free_params_cb(tse_task_t *task, void *data)
 	return rc;
 }
 
+/* Call-back that recovers EC allocated memory for fetch  */
+static int
+ec_free_fetch_params_cb(tse_task_t *task, void *data)
+{
+	struct ec_fetch_params *head = *((struct ec_fetch_params **)data);
+	int			rc = task->dt_result;
+
+	ec_free_fetch_params(head);
+	return rc;
+}
+
+
 /* Identifies the applicable subset of forwarding targets for non-full-stripe
- * EC updates.
+ * EC updates. If called for EC fetch, the tgt_set is set to the addressed data
+ * targets.
+ *
+ * For single values, the tgt_set includes the first data target, and all parity
+ * targets for update. For fetch, the first data target is selected. (This will
+ * change once encoding of single values is supported).
  */
 void
 ec_get_tgt_set(daos_iod_t *iods, unsigned int nr, struct daos_oclass_attr *oca,
 	       bool parity_include, uint64_t *tgt_set)
 {
 	unsigned int    len = oca->u.ec.e_len;
 	unsigned int    k = oca->u.ec.e_k;
-	unsigned int    p = oca->u.ec.e_p;
-	uint64_t	ss = k * len;
-	uint64_t	par_only = (1UL << p) - 1;
-	uint64_t	full;
+	unsigned int    p_offset, p = oca->u.ec.e_p;
+	uint64_t	ss;
+	uint64_t	full = (1UL << (k+p)) - 1;
 	unsigned int	i, j;
 
-	if (parity_include) {
-		full = (1UL << (k+p)) - 1;
+	if (parity_include)
 		for (i = 0; i < p; i++)
 			*tgt_set |= 1UL << i;
-	} else {
-		full = (1UL << k) - 1;
-	}
+
 	for (i = 0; i < nr; i++) {
-		if (iods->iod_type != DAOS_IOD_ARRAY)
+		if (iods->iod_type != DAOS_IOD_ARRAY) {
+			*tgt_set |= 1UL << p;
 			continue;
+		}
+
+		if (!parity_include)
+			full = ((1UL << (k+p)) - 1) - ((1UL << p) - 1);
```

Comment:
if we have already computed "full" at L454, how about we just check parity_included at there as well?

Reply:
I've changed this so full is all targets for update (parity include), and all data targets for fetch.  this function was initially coded to support update (and all updates involve the parity targets). When fetch was added, the parity_include boolean was added to the signature.